File: Baekjoon/11066파일합치기.cpp
#include<iostream>
#include<algorithm>
#define INFINITY 987654321
using namespace std;
int n;
int arr[501];
int sum[501];
int cache[501][501];

int unionFile(int start, int end) {
	if (cache[start][end] != INFINITY) {
		return cache[start][end];
	}
	else if (start == end) {
		return cache[start][end] = 0;
	}
	else if (start + 1 == end) {
		return cache[start][end] = arr[start] + arr[end];
	}
	for (int mid = start; mid < end; mid++) {
		cache[start][end] = min(cache[start][end], unionFile(start, mid) + unionFile(mid + 1, end));
	}
	return cache[start][end] += sum[end] - sum[start - 1];
}

int main() {
	int testcase;
	cin >> testcase;
	while (testcase--) {
		cin >> n;
		fill(&cache[0][0], &cache[n][n] + 1, INFINITY);
		for (int i = 1; i <= n; i++) {
			cin >> arr[i];
			sum[i] = arr[i] + sum[i - 1];
		}
		cout << unionFile(1, n) << '\n';
	}

	return 0;
}

File: Baekjoon/11066파일합치기.py
from copy import deepcopy
INF = 10 ** 9

def solution(start, end):
    if dp[start][end] != INF:
        return dp[start][end]
    elif start == end:
        dp[start][end] = 0
        return 0
    elif start + 1 == end:
        dp[start][end] = arr[start] + arr[end]
        return dp[start][end]

    for mid in range(start, end):
        dp[start][end] = min(dp[start][end], solution(start, mid) + solution(mid + 1, end))

    dp[start][end] += prefixSum[end] - prefixSum[start - 1]
    return dp[start][end]

t = int(input())
while t:
    t -= 1
    n = int(input())
    arr = [0] + list(map(int, input().split()))
    prefixSum = deepcopy(arr)
    for i in range(1, n + 1):
        prefixSum[i] += prefixSum[i - 1]
    dp = [[INF] * (n + 1) for _ in range(n + 1)]
    print(solution(1, n))